compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangTernaryExpr.java

```diff
@@ -33,6 +34,8 @@ public class BLangTernaryExpr extends BLangExpression implements TernaryExpressi
     public BLangExpression expr;
     public BLangExpression thenExpr;
     public BLangExpression elseExpr;
+    public Scope thenScope;
+    public Scope elseScope;
```

Comment:
IMO this is not the correct fix. Normally we introduce scope to the symbol to keep its child symbols. 

```ballerina
function test() {
  int? i = 3;
  int|int[] thenResult = i is int ?
      from var _ in [1, 2]
      where i + 2 == 5
      select 2 : 2;
}
```
When considering the above scenario, we use the variable `i` within the lambda function created for the `where` clause. So, in the query desugaring logic, it is needed to identify `i` as a closure variable. But with the current approach we didn't mark `i` as a closure variable in the query desugaring logic. Due to this above program fails in the runtime.

The reason for not to identify `i` as closure variable(in query desugaring phase) is in the type narrowing phase we generate new symbol for `i`. But within the new symbol which generated for `i` we keep the `originalsymbol` as well. We can use that symbol to check whether that the `i` is closure variable or not.

Solution for this is to use the `originalsymbol` instead of using the current symbol to check whether it is a closure variable or not when visiting the variable reference in the query desugaring phase. 

Example
```java
if (symbol instanceof BVarSymbol) {
    BVarSymbol originalSymbol = ((BVarSymbol) symbol).originalSymbol;
    if (originalSymbol != null) {
        symbol = originalSymbol;
    }
}
```

Reply:
Although this solves the problem with the ternary case, the same error now occurs in the if-body case (which was passed with the current implementation).

```ballerina
public function main() {
    testTernaryWithQuery();
}

function testTernaryWithQuery() {
    int? i = 3;

    int|int[] thenResult;
    if i is int {
        thenResult = from var _ in [1, 2]
        where i + 2 == 5
        select 2;
    } else {
        thenResult = 2;
    }
}
```

Since the narrowed environment scope is properly defined for the `if` and `else` bodies, it will return the narrowed symbol rather than the original symbol when invoking `lookupClosureVarSymbol`. As a result, the proposed approach will not identify `i` as a closure since the desugaring symbol and the resolved symbol are now different.

Reply:
The var-ref logic in query desugar is different. Basically it gets the symbol from node and symbol from env and compare. If we change the logic to consider `originalSymbol` as you mentioned, `ifStmt` case can be failed. I don't think it is feasible to do a whole refactoring for var-ref logic in query desugar.

Reply:
As discussed offline, I also don't think we should keep the scopes in the node. Let's check if there's a different way to identify and mark captured variables correctly.